### ISSUE TITLE
Support for .ecrc.json

### ISSUE THREAD

**Issue by nikolay** [editorconfig-checker/editorconfig-checker#306]
Please fall back to `.ecrc.json` configuration file if `.ecrc` does not exist. It's much easier to validate JSON against the `.json` extension and not add many extra one-off files to a project. Many other tools started to support such options to ease config file validation.

**Comment by polarathene** [comment]
I think a common convention for `.*rc` files which look like JSON is to be JSONC, which is basically JSON with comments supported (_sometimes also with support for trailing commas_).

Other than that, it's still JSON.

**Comment by zoechi** [comment]
I'd prefer yaml because then I also could add comments
I think `.ecrc` is a pretty bad name. Nobody onboarding to a project will understand what that file does and I'm not able to add a comment that hints to the tool it belongs to.
I think the name should be `editorconfig-checker.json` or `editorconfig-checker.yaml`

**Comment by polarathene** [comment]
> I'd prefer yaml because then I also could add comments

JSONC allows comments, that's what you get with `.ecrc` no?

> Nobody onboarding to a project will understand what that file does and I'm not able to add a comment that hints to the tool it belongs to.

They don't need to understand it? How often is that file being modified? It's a file for linting rules, you may have more than one related to that, have a document that centralizes that information all in one place, direct the new hire to that. Easy to lookup when needed.

You can call the file whatever you like though. This is [how a project I work on uses it](https://github.com/docker-mailserver/docker-mailserver/blob/25c7024cc4c7a6ee81be70144f6ecaf4fddf44ca/test/linting/lint.sh#L20-L27):

```bash
docker run --rm --tty \
    --volume "${REPOSITORY_ROOT}:/check:ro" \
    --workdir "/check" \
    --name dms-test_eclint \
    "mstruebing/editorconfig-checker:${ECLINT_VERSION}" ec -config "/check/test/linting/.ecrc.json"
```

Pretty obvious what tool is being used there, and the `-config` option let's you point to any filename.

**Comment by nikolay** [comment]
@zoechi YAML or JSON would be nice as long as we don't have to add yet another exception for a config file, which format must be hardcoded. We all use automated tools, CI, etc. It's not smart to keep adding filenames to automatic formatting scripts instead of `**/*.json`, let's say. I agree - `.ecrc` is not self-explanatory. The issue with YAML is that some people use `.YML`, others -. `YAML`, so, literally, you need to fall back to both.

**Comment by nikolay** [comment]
@polarathene I'm fully aware of this and already using it in scripts. But sometimes I run manually, and I always have to add the config option.

**Comment by per1234** [comment]
Hi all. This discussion is getting off topic. The proposal is to support the filename `.ecrc.json` for the EditorConfig configuration file. If you have some original and valuable information to add specific to that proposal, you are welcome to comment, but anything else is not productive.

If you would like to propose other enhancements such as support for YAML in configuration files, support for JSONC in the configuration file, support for more descriptive configuration filenames, etc. please feel free to create a dedicated issue for those proposals.

**Comment by jerone** [comment]
There is the [widely used and popular](https://www.npmjs.com/browse/depended/cosmiconfig) package called [`cosmiconfig`](https://github.com/cosmiconfig/cosmiconfig) that supports many formats, including the `rc`, `.json`, `.yaml`, and more.

It also supports the `{module}.config.js` format (_my personal preference_), that is getting a lot of traction lately (e.g. ESLint). 
The latest version also checks for config files in the `.config` folder, allowing for a more cleaner and reorganized repo.
These two are again user preferences, but with `cosmiconfig`, you don't have to deal with it, and it becomes an user decision.

**Comment by ccoVeille** [comment]
There is a current discussion in progress in the refactoring:
- #371

> What about `.ecrc`? Should the configuration file named `.editorconfig-checkerrc`? (might be a separate issue, but as it's a JSON file, we should support the file with the extension, so `.editorconfig-checkerrc.json`).



_Originally posted by @theoludwig in https://github.com/editorconfig-checker/editorconfig-checker/pull/371#pullrequestreview-2287182453_

Let's continue the discussion here.

I do think the .ecrc file should be superseded by a file named 
`.editorconfig-checker.json`

The current name is cryptic. People don't see the link between this file and the tool.

That said there are many projects using this file. It means that if there is a change to do, the code should look for the new name first, then fallbacks to .ecrc file if missing

About the format, I would say while it might be interesting to have different formats (.yaml, .toml …), it should be addressed in a separate issue as the code change would be bigger than simply looking for a different filenames.

I think #371 is the good moment to introduce a possible new filename with `.editorconfig-checker.json` in addition of `.ecrc `

**Comment by klaernie** [comment]
I too think that releasing a rename of `.ecrc` together with the changes in #371 is the right thing to do.

I also think the current name is really cryptic, and possibly some internals should be adjusted. I also dislike that the name is missing it's file extension - it should be named according to it's content, so people do not think it is a yaml file. And once we've switched to a new name writing a schema for #194 would be nice too. 

Regarding formats I'm a big on the fence - maybe the biggest benefit would be switching to JSON5 to allow bare word keys and trailing commas in arrays and objects while being completely compatible with the current JSON format syntax wise. YAML would be identical in the support for the features we need, while being easier to write as well. I'm not too much a fan of TOML, I dislike it's structure when writing more complex objects. At work we've coalesced on YAML for configuration that humans need to write, but here we have a precedent for the JSON family.  

In the end supporting multiple formats would be overkill for the current use case, where consistency is king. I would like to prevent the authors of linter collections like super-linter or mega-linter having to code complex situations to decide if their opinionated config file should be applied or if the user provided a valid configuration already.

So to sum it up: my vote would be for `.editorconfig-checker.json5`, falling back to `.ecrc` while printing a deprecation notice when doing so.

**Comment by ccoVeille** [comment]
I'm fine with the .json5 extension. It's a good thing to use it

It's a good idea to allow comments and trailing commas.

Also YAML is a superset of JSON, so supporting both should be easy if there is a move to YAML (famous last words :tm: )

I'm fine with having only JSON5 and YAML, toll and INI are a pain when you need to write configuration with multiple levels
